File: April 30 Day Challenge/Week 3 April 15th - April 21st/Minimum Path Sum.cpp
/*
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Example:

Input:
[
  [1,3,1],
  [1,5,1],
  [4,2,1]
]
Output: 7
Explanation: Because the path 1→3→1→1→1 minimizes the sum.

*/
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        if(grid.size() == 0) return 0;
        
        int N = grid.size();
        int M = grid[0].size();
        
        vector<vector<int>> dp(N, vector<int>(M, 0));
        
        for(int i=0;i<N; i++){
            for(int j=0; j<M; j++){
                
                if(i == 0 && j == 0){
                    dp[i][j] = grid[i][j];
                    continue;
                }
                else if(i==0){
                    dp[i][j] = dp[i][j-1] + grid[i][j];
                    continue;
                }
                else if(j==0){
                    dp[i][j] = dp[i-1][j] + grid[i][j];
                    continue;
                }
                else{
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j];
                }
                
            }
        }
        return dp[N-1][M-1];
    }
};

File: Medium/Minimum Path Sum.java
/*
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Example:

Input:
[
  [1,3,1],
  [1,5,1],
  [4,2,1]
]
Output: 7
Explanation: Because the path 1→3→1→1→1 minimizes the sum.
*/

class Solution {
    public int minPathSum(int[][] grid) {
        if(grid == null || grid.length == 0){
            return 0;
        }
        int rows = grid.length, columns = grid[0].length;
        Integer[][] dp = new Integer[rows][columns];
        
        for(Integer row = 0; row < rows; ++row) {
            for(Integer column = 0; column < columns; ++column) {
                if(row == 0 && column == 0) {
                    dp[row][column] = grid[0][0];
                    continue;
                }
                if(row == 0) {
                    dp[row][column] = dp[row][column - 1] + grid[row][column];
                }
                else if(column == 0) {
                    dp[row][column] = dp[row - 1][column] + grid[row][column];
                }
                else {
                    dp[row][column] = Math.min(dp[row - 1][column], dp[row][column - 1]) + grid[row][column];
                }
            
            }
        }
        return dp[rows - 1][columns - 1];
    }
}